// Copyright 2019, Beeri 15.  All rights reserved.
// Author: Roman Gershman (romange@gmail.com)
//
syntax = "proto2";

package mr3.pb;

message WireFormat {
  enum Type {
    LST = 2;
    TXT = 3;
  }
  required Type type = 1;
}

message ShardSpec {
  enum Type {
    MODN = 1;
    USER_DEFINED = 2;
  }

  required Type type = 1;
  optional uint32 modn = 2;
}

message Input {
  required string name = 1;

  required WireFormat format = 2;

  message FileSpec {
    required string url_glob = 1;
    optional string metadata = 2;

    oneof shard_id_ref {
      string custom_shard_id = 3;
      uint32 shard_id = 4;
   }
  };

  // In case of sharded input, each file_spec corresponds to a shard.
  repeated FileSpec file_spec = 4;
}

message Output {
  required string name = 1;
  required WireFormat format = 2;

  enum CompressType {
    NONE = 0;
    GZIP = 1;
  }

  message Compress {
    required CompressType type = 1;
    optional int32 level = 2 [default = 1];
  }

  optional Compress compress = 3;
  optional ShardSpec shard_spec = 4;
}


// Can be mapper or joiner.
message Operator {
  repeated string input_name = 1;  // corresponds to the name in Input.name.

  optional string op_name = 2;

  required Output output = 3;

  enum Type {
    MAP = 1;
    HASH_JOIN = 2;
  }
  optional Type type = 4;
}

message Pipeline {
  map<string, Input> input = 1;

  repeated Operator operator = 2;
}
